Use system background color for wide-layout attachments on macOS

squelart opened the pull request:
#### 2dba621b4a765858360e25e481a4d85ba42f0bf8
<pre>
Use system background color for wide-layout attachments on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=255691">https://bugs.webkit.org/show_bug.cgi?id=255691</a>
rdar://105252500

Reviewed by Tim Horton and Aditya Keerthi.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/mac/NSColorSPI.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserIdioms.cpp:
(WebCore::isValueAllowedInMode):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):

Canonical link: <a href="https://commits.webkit.org/263211@main">https://commits.webkit.org/263211@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85c9fa7c53f4a0e9397fd019bd750a025ba8222f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3950 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5389 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4202 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4131 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4020 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4097 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3998 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4188 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5270 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1690 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3538 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5603 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3255 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5012 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3726 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4001 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4030 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3538 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/997 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/962 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3568 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4123 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3798 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1109 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -34,7 +34,7 @@ div#attachment-container {
     width: 266px;
     height: 80px;
-    background-color: rgb(242, 242, 242); /* FIXME: Use quinary-fill when available, see rdar://105252500 depending on rdar://104844460 */
+    background-color: -apple-system-tertiary-fill;
```
Note to reviewers: The FIXME was incorrect about "quinary-fill" (due to a mismatch in the spec author's palette names!), tertiary-fill is actually what we want.

hortont424 commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -716,6 +716,11 @@ static Color activeButtonTextColor()
+        case CSSValueAppleSystemTertiaryFill:
+            if ([NSColor respondsToSelector:@selector(tertiarySystemFillColor)])
+                return semanticColorFromNSColor([NSColor tertiarySystemFillColor]);
+            return semanticColorFromNSColor([NSColor colorWithGenericGamma22White:0.0 alpha:0.05]);
```
Why does this one need to bounce through NSColor? Can it be more like the others above?

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSColorSPI.h:
```diff
@@ -63,3 +63,7 @@ typedef NS_ENUM(NSInteger, NSUserAccentColor) {
 
 #endif
+
+@interface NSColor (Staging_105252500)
```
This should use the radar number where the SPI was added, not your radar.

pxlcoder replied:
Need a `FIXME` plus radar to remove the staging code.

pxlcoder replied:
This needs it's own `HAVE()` macro since it's not available on older OSes.

pxlcoder commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -305,6 +305,7 @@ accentcolortext
 -apple-system-selected-text-background enable-if=WTF_PLATFORM_MAC
 -apple-system-unemphasized-selected-text-background enable-if=WTF_PLATFORM_MAC
 -apple-system-find-highlight-background enable-if=WTF_PLATFORM_MAC
+-apple-system-tertiary-fill enable-if=WTF_PLATFORM_MAC
```
We need to restrict this to UA sheet only (similar to the color you added for iOS).

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -716,6 +716,11 @@ static Color activeButtonTextColor()
 
+        case CSSValueAppleSystemTertiaryFill:
+            if ([NSColor respondsToSelector:@selector(tertiarySystemFillColor)])
+                return semanticColorFromNSColor([NSColor tertiarySystemFillColor]);
```
This could go in the selector list above.

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -716,6 +724,12 @@ static Color activeButtonTextColor()
         }
 
+        // FIXME: Remove this fallback when AppKit without tertiary-fill is not used anymore; see rdar://108340604.
+        case CSSValueAppleSystemTertiaryFill:
+            if (localAppearance.usingDarkAppearance())
+                return { SRGBA<uint8_t> { 255, 255, 255, 13 }, Color::Flags::Semantic };
+            return { SRGBA<uint8_t> { 0, 0, 0, 13 }, Color::Flags::Semantic };
```
Let's put this in an `#if !HAVE(NSCOLOR_FILL_COLOR_HIERARCHY)`.

squelart replied:
I was thinking that the fallback could *always* be there, because the css value doesn't have a build condition -- should it? (And is it possible? I didn't see any in CSSValueKeywords.in. 🤔)

squelart replied:
But I'll add the `#if` anyway; let me know if I should remove it after all...

pxlcoder replied:
The build condition in CSSValueKeywords.in, is `PLATFORM(MAC)`.

But you're right that we should probably always have the fallback without the `#if` until the staging code is removed (which can be done in a couple weeks), to keep tests passing. The `HAVE` is going to be around for longer than that.

pxlcoder commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1459,3 +1459,7 @@
 #endif
+
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000)
+#define HAVE_NSCOLOR_FILL_COLOR_HIERARCHY 1
```
I would call this `HAVE_NSCOLOR_TERTIARY_SYSTEM_FILL_COLOR`.

squelart replied:
Well, why not allow future users of other fill colors to be able to reuse this definition? 😉 
(All fill colors were added together in rdar://104764768 (Add "fill" color hierarchy to NSColor).)

pxlcoder replied:
That's a good point! We can leave as-is.

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -716,6 +724,14 @@ static Color activeButtonTextColor()
             return semanticColorFromNSColor(alternateColors[1]);
         }
 
+#if HAVE(NSCOLOR_FILL_COLOR_HIERARCHY)
```
This should have been `!HAVE`, but I think we can just remove now, based on the discussion above.

squelart committed:
Use system background color for wide-layout attachments on macOS
https://bugs.webkit.org/show_bug.cgi?id=255691
rdar://105252500

Reviewed by Tim Horton and Aditya Keerthi.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/mac/NSColorSPI.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserIdioms.cpp:
(WebCore::isValueAllowedInMode):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):

Canonical link: https://commits.webkit.org/263211@main
